cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1224,75 +1224,30 @@ func generateMountsModeFactoryReset(mst *initramfsMountsState) error {
 	// will be wiped anyway so we do not even bother looking up those
 	// partitions (which may be corrupted too, hence factory-reset was
 	// invoked)
-	degradedState := recoverDegradedState{}
-	part := degradedState.partition("ubuntu-save")
-	degraded := false
-	isEncrypted := false
-	_, findErr := disk.FindMatchingPartitionWithFsLabel(secboot.EncryptedPartitionName("ubuntu-save"))
-	if findErr == nil {
-		// found it and it's encrypted
-		isEncrypted = true
-		unlockOpts := &secboot.UnlockVolumeUsingSealedKeyOptions{
-			// we want to allow using the recovery key if the fallback key fails
-			AllowRecoveryKey: true,
-			WhichModel:       func() (*asserts.Model, error) { return model, nil },
-		}
-		saveFallbackKey := filepath.Join(boot.InitramfsSeedEncryptionKeyDir, "ubuntu-save.recovery.sealed-key")
-		// TODO: this prompts again for a recover key, but really this is the
-		// reinstall key we will prompt for
-		// TODO: we should somehow customize the prompt to mention what key we need
-		// the user to enter, and what we are unlocking (as currently the prompt
-		// says "recovery key" and the partition UUID for what is being unlocked)
-		unlockRes, unlockErr := secbootUnlockVolumeUsingSealedKeyIfEncrypted(disk, "ubuntu-save", saveFallbackKey, unlockOpts)
-		part.FindState = partitionFound
-		part.partDevice = unlockRes.PartDevice
-		part.fsDevice = unlockRes.FsDevice
-		if unlockErr != nil {
-			degradedState.LogErrorf("cannot unlock encrypted ubuntu-save (device %v) with sealed fallback key: %v", part.partDevice, unlockErr)
-			part.UnlockState = partitionErrUnlocking
-			degraded = true
-		} else {
-			// unlocked successfully
-			part.UnlockState = partitionUnlocked
-			switch unlockRes.UnlockMethod {
-			case secboot.UnlockedWithSealedKey:
-				part.UnlockKey = keyFallback
-			case secboot.UnlockedWithRecoveryKey:
-				part.UnlockKey = keyRecovery
-				degraded = true
+	machine, err := func() (machine *recoverModeStateMachine, err error) {
+		allowFallback := true
+		machine = newRecoverModeStateMachine(model, disk, allowFallback)
+		// start from looking up encrypted ubuntu-save and unlocking with the fallback key
+		machine.current = machine.unlockMaybeEncryptedAloneSaveFallbackKey
+		for {
```

Comment:
nice this is very clean, I hadn't realized that all the same logic is actually true if we set ourselves to this state, but looking at the state diagram this seems correct